test/unit/Formatter/JsonFormatterTest.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+declare(strict_types=1);
+
+namespace RoaveTest\BackwardCompatibility\Formatter;
+
+use PHPUnit\Framework\TestCase;
+use Psl\Env;
+use Psl\Filesystem;
+use ReflectionException;
+use Roave\BackwardCompatibility\Change;
+use Roave\BackwardCompatibility\Changes;
+use Roave\BackwardCompatibility\Formatter\JsonFormatter;
+use Roave\BackwardCompatibility\Git\CheckedOutRepository;
+use Symfony\Component\Console\Output\BufferedOutput;
+use function Psl\Filesystem;
+
+/** @covers \Roave\BackwardCompatibility\Formatter\JsonFormatter */
+final class JsonFormatterTest extends TestCase
+{
+    /** @throws ReflectionException */
+    public function testWrite(): void
+    {
+        $output            = new BufferedOutput();
+        $temporaryLocation = Filesystem\create_temporary_file(Env\temp_dir(), 'jsonFormatter');
+
+        Filesystem\delete_file($temporaryLocation);
+        Filesystem\create_directory($temporaryLocation . '/foo/bar/.git');
+
+        (new JsonFormatter(
+            $output,
+            CheckedOutRepository::fromPath($temporaryLocation . '/foo/bar'),
+        ))->write(Changes::fromList(
+            Change::removed('foo', true),
+            Change::added('bar', false),
+            Change::changed('baz', false)
+                ->onFile('baz-file.php'),
+            Change::changed('tab', false)
+                ->onFile('tab-file.php')
+                ->onLine(5),
+            Change::changed('taz', false)
+                ->onFile('taz-file.php')
+                ->onLine(6)
+                ->onColumn(15),
+            Change::changed('tar', false)
+                ->onFile('tar-file.php')
+                ->onLine(-1)
+                ->onColumn(-1),
+            Change::changed('file-in-checked-out-dir', false)
+                ->onFile($temporaryLocation . '/foo/bar/subpath/file-in-checked-out-dir.php')
+                ->onLine(10)
+                ->onColumn(20),
+        ));
+
+        Filesystem\delete_directory($temporaryLocation, true);
+
+        $expected = [
+            'errors' => [
+                ['description'=>'foo','path'=>null,'line'=>null,'column'=>null],
+                ['description'=>'bar','path'=>null,'line'=>null,'column'=>null],
+                ['description'=>'baz','path'=>'baz-file.php','line'=>null,'column'=>null],
+                ['description'=>'tab','path'=>'tab-file.php','line'=>5,'column'=>null],
+                ['description'=>'taz','path'=>'taz-file.php','line'=>6,'column'=>15],
+                ['description'=>'tar','path'=>'tar-file.php','line'=>-1,'column'=>-1],
+                ['description'=>'file-in-checked-out-dir','path'=>'subpath/file-in-checked-out-dir.php','line'=>10,'column'=>20],
+            ]
+        ];
+
+        $result = $output->fetch();
+        $data = json_decode(trim($result), true);
+        $this->assertIsArray($data);
+        $this->assertEquals($expected, $data);
+
+        self::assertEquals(
```

Comment:
Possibly worth using PHPUnit's built-in Json assertions here: it gives much better error messages, if anything shifts for whatever reason